sui-execution/latest/sui-adapter/src/programmable_transactions/trace_utils.rs

```diff
@@ -0,0 +1,246 @@
+    context: &mut ExecutionContext<'_, '_, '_>,
+    trace_builder_opt: &mut Option<MoveTraceBuilder>,
+    obj_values: &[ObjectValue],
+) -> Result<(), ExecutionError> {
+    if let Some(trace_builder) = trace_builder_opt {
+        let to_transfer = obj_values
+            .into_iter()
+            .map(|v| obj_info_from_obj_value(context, v))
+            .collect::<Result<Vec<_>, _>>()?;
+        trace_builder.push_event(TraceEvent::External(Box::new(serde_json::json!(
+            PTBExternalEvent::TransferObjects(TransferEvent { to_transfer })
+        ))));
+    }
+    Ok(())
+}
+
+/// Inserts PTB summary event into the trace.
+pub fn trace_ptb_summary(
+    trace_builder_opt: &mut Option<MoveTraceBuilder>,
+    commands: &[Command],
+) -> Result<(), ExecutionError> {
+    if let Some(trace_builder) = trace_builder_opt {
+        trace_builder.push_event(TraceEvent::External(Box::new(serde_json::json!(
+            PTBExternalEvent::Summary(
+                commands
+                    .iter()
+                    .map(|c| match c {
+                        Command::MoveCall(move_call) => {
+                            let pkg = move_call.package.to_string();
+                            let module = move_call.module.clone();
+                            let function = move_call.function.clone();
+                            PTBCommandInfo::MoveCall {
+                                pkg,
+                                module,
+                                function,
+                            }
+                        }
+                        Command::TransferObjects(..) => PTBCommandInfo::TransferObjects,
+                        Command::SplitCoins(..) => PTBCommandInfo::SplitCoins,
+                        Command::MergeCoins(..) => PTBCommandInfo::MergeCoins,
+                        Command::Publish(..) => PTBCommandInfo::Publish,
+                        Command::MakeMoveVec(..) => PTBCommandInfo::MakeMoveVec,
+                        Command::Upgrade(..) => PTBCommandInfo::Upgrade,
+                    })
+                    .collect(),
+            )
+        ))));
+    }
+
+    Ok(())
+}
+
+/// Inserts split coins event into the trace.
+pub fn trace_split_coins(
+    context: &mut ExecutionContext<'_, '_, '_>,
+    trace_builder_opt: &mut Option<MoveTraceBuilder>,
+    coin_type: &Type,
+    input_coin: &Coin,
+    split_coin_values: &Vec<Value>,
+) -> Result<(), ExecutionError> {
+    if let Some(trace_builder) = trace_builder_opt {
+        let type_tag_with_refs = trace_type_to_type_tag_with_refs(context, coin_type)?;
+        let split_coin_infos = split_coin_values
+            .iter()
+            .map(|coin_val| {
+                let Value::Object(ObjectValue {
+                    contents: ObjectContents::Coin(split_coin),
+                    ..
+                }) = coin_val
+                else {
+                    invariant_violation!("Expected result of split coins PTB command to be a coin");
```

Comment:
Should we introduce new types of execution errors instead of using invariant violations to report tracing-related errors?